src/OpenTelemetry.Contrib.Instrumentation.MySqlData/MySqlDataInstrumentation.cs

```diff
@@ -0,0 +1,225 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using MySql.Data.MySqlClient;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.MySqlData
+{
+    /// <summary>
+    /// Mysql.Data instrumentation.
+    /// </summary>
+    internal class MySqlDataInstrumentation : DefaultTraceListener
+    {
+        private readonly ConcurrentDictionary<long, MySqlConnectionStringBuilder> dbConn = new ConcurrentDictionary<long, MySqlConnectionStringBuilder>();
+
+        private readonly MySqlDataInstrumentationOptions options;
+
+        public MySqlDataInstrumentation(MySqlDataInstrumentationOptions options = null)
+        {
+            this.options = options ?? new MySqlDataInstrumentationOptions();
+            MySqlTrace.Listeners.Clear();
+            MySqlTrace.Listeners.Add(this);
+            MySqlTrace.Switch.Level = SourceLevels.Information;
+            MySqlTrace.QueryAnalysisEnabled = true;
+        }
+
+        /// <inheritdoc />
+        public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType eventType, int id, string format, params object[] args)
+        {
+            try
+            {
+                switch ((MySqlTraceEventType)id)
+                {
+                    case MySqlTraceEventType.ConnectionOpened:
+                        // args: [driverId, connStr, threadId]
+                        var driverId = (long)args[0];
+                        var connStr = args[1].ToString();
+                        this.dbConn[driverId] = new MySqlConnectionStringBuilder(connStr);
+                        break;
+                    case MySqlTraceEventType.ConnectionClosed:
+                        break;
+                    case MySqlTraceEventType.QueryOpened:
+                        // args: [driverId, threadId, cmdText]
+                        this.BeforeExecuteCommand(this.GetCommand(args[0], args[2]));
+                        break;
+                    case MySqlTraceEventType.ResultOpened:
+                        break;
+                    case MySqlTraceEventType.ResultClosed:
+                        break;
+                    case MySqlTraceEventType.QueryClosed:
+                        // args: [driverId]
+                        this.AfterExecuteCommand();
+                        break;
+                    case MySqlTraceEventType.StatementPrepared:
+                        break;
+                    case MySqlTraceEventType.StatementExecuted:
+                        break;
+                    case MySqlTraceEventType.StatementClosed:
+                        break;
+                    case MySqlTraceEventType.NonQuery:
+                        break;
```

Comment:
I'm not familiar with the functionality of the `MySqlTrace` class, so I don't know what to expect from these different event types. This `NonQuery` event type makes me wonder if inserts/deletes/updates would not be covered by this instrumentation. Do you know if QueryOpened/Closed covers these use cases?

Reply:
The `NonQuery` event only triggered when `SetDatabase` called, QueryOpened/Closed will cover select/insert/update/delete.

reference:
https://github.com/mysql/mysql-connector-net/blob/e39b64edbb7d2b44ffd05b2ab36a399bed8f24e8/MySQL.Data/src/TracingDriver.cs#L140-L145